### ISSUE TITLE
Libtation requests unavailable quality

### ISSUE THREAD

**Issue by maaximal** [rmcrackan/Libation#318]
German audible offers audiobooks which are free to listen to while a premium membership is running. These can be added to the library and I was able to download some of them back in June. Now when I try to download them it seems like audible only offers "High" or "Normal" audio quality for them but libtation tries to get the "Extreme" audio version:
```
 {}2022-07-19 19:40:45.897 +02:00 [ERR] (at AudibleApi.Api+<GetDownloadLicenseAsync>d__19.MoveNext()) Error requesting download license
AudibleApi.ApiErrorException: Error requesting license for asin: [B07SQF3FQ9]
 ---> AudibleApi.ValidationErrorException: 1 validation error detected: Value 'Extreme' at 'quality' failed to satisfy constraint: Member must satisfy enum value set: [High, Normal]
   at AudibleApi.RestMessageValidator.ThrowStrongExceptionsIfInvalid(String message, Uri requestUri) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/RestMessageValidator.cs:line 34
   at AudibleApi.ApiMessageHandler.ProcessResponse(HttpResponseMessage response, CancellationToken cancellationToken) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/ApiMessageHandler.cs:line 69
   at System.Net.Http.MessageProcessingHandler.<>c.<SendAsync>b__5_0(Task`1 task, Object state)
--- End of stack trace from previous location ---
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at AudibleApi.ApiUnauthenticated.SendClientRequest(IHttpClientActions client, HttpRequestMessage request) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/ApiUnauthenticated [partial].cs:line 51
   at AudibleApi.Api.AdHocAuthenticatedRequestAsync(String requestUri, HttpMethod method, IHttpClientActions client, JObject postData) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/Api [partial].cs:line 57
   at AudibleApi.Api.GetDownloadLicenseAsync(String asin, DownloadQuality quality) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/Api.Download.cs:line 49
   --- End of inner exception stack trace ---
 {"ExceptionDetail":{"HResult":-2146233088,"Message":"Error requesting license for asin: [B07SQF3FQ9]","Source":null,"InnerException":{"HResult":-2146233088,"Message":"1 validation error detected: Value 'Extreme' at 'quality' failed to satisfy constraint: Member must satisfy enum value set: [High, Normal]","Source":"AudibleApi","TargetSite":"Void ThrowStrongExceptionsIfInvalid(System.String, System.Uri)","RequestUri":"https://api.audible.de/1.0/content/B07SQF3FQ9/licenserequest","JsonMessage":"{\"message\":\"1 validation error detected: Value 'Extreme' at 'quality' failed to satisfy constraint: Member must satisfy enum value set: [High, Normal]\"}","Type":"AudibleApi.ValidationErrorException"},"RequestUri":"/1.0/content/B07SQF3FQ9/licenserequest","JsonMessage":"{\"consumption_type\":\"Download\",\"supported_drm_types\":[\"Adrm\",\"Mpeg\"],\"quality\":\"Extreme\",\"response_groups\":\"last_position_heard,pdf_url,content_reference,chapter_info\"}","Type":"AudibleApi.ApiErrorException"}}2022-07-19 19:40:45.899 +02:00 [INF] (at LibationWinForms.ProcessQueue.LogMe+<>c.<.ctor>b__9_0(System.Object, System.String)) Automated backup: Download & Decrypt Step, Completed: [B07SQF3FQ9] Der Dieb und der Söldner: Die Gaukler-Chroniken 1
 {}2022-07-19 19:40:45.903 +02:00 [ERR] (at LibationWinForms.ProcessQueue.LogMe+<>c.<.ctor>b__9_2(System.Object, System.ValueTuple`2[[System.Exception, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.String, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]])) Download & Decrypt
AudibleApi.ApiErrorException: Error requesting license for asin: [B07SQF3FQ9]
 ---> AudibleApi.ValidationErrorException: 1 validation error detected: Value 'Extreme' at 'quality' failed to satisfy constraint: Member must satisfy enum value set: [High, Normal]
   at AudibleApi.RestMessageValidator.ThrowStrongExceptionsIfInvalid(String message, Uri requestUri) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/RestMessageValidator.cs:line 34
   at AudibleApi.ApiMessageHandler.ProcessResponse(HttpResponseMessage response, CancellationToken cancellationToken) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/ApiMessageHandler.cs:line 69
   at System.Net.Http.MessageProcessingHandler.<>c.<SendAsync>b__5_0(Task`1 task, Object state)
--- End of stack trace from previous location ---
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at AudibleApi.ApiUnauthenticated.SendClientRequest(IHttpClientActions client, HttpRequestMessage request) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/ApiUnauthenticated [partial].cs:line 51
   at AudibleApi.Api.AdHocAuthenticatedRequestAsync(String requestUri, HttpMethod method, IHttpClientActions client, JObject postData) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/Api [partial].cs:line 57
   at AudibleApi.Api.GetDownloadLicenseAsync(String asin, DownloadQuality quality) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/Api.Download.cs:line 49
   --- End of inner exception stack trace ---
   at AudibleApi.Api.GetDownloadLicenseAsync(String asin, DownloadQuality quality) in /home/runner/work/AudibleApi/AudibleApi/AudibleApi/Api.Download.cs:line 65
   at FileLiberator.DownloadDecryptBook.downloadAudiobookAsync(LibraryBook libraryBook) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2022\Libation\Source\FileLiberator\DownloadDecryptBook.cs:line 103
   at FileLiberator.DownloadDecryptBook.ProcessAsync(LibraryBook libraryBook) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2022\Libation\Source\FileLiberator\DownloadDecryptBook.cs:line 56
   at FileLiberator.Processable.ProcessSingleAsync(LibraryBook libraryBook, Boolean validate) in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2022\Libation\Source\FileLiberator\Processable.cs:line 53
   at LibationWinForms.ProcessQueue.ProcessBook.ProcessOneAsync() in C:\Dropbox\DinahsFolder\coding\_NET\Visual Studio 2022\Libation\Source\LibationWinForms\ProcessQueue\ProcessBook.cs:line 103
 {"ExceptionDetail":{"HResult":-2146233088,"Message":"Error requesting license for asin: [B07SQF3FQ9]","Source":"AudibleApi","TargetSite":"Void MoveNext()","InnerException":{"HResult":-2146233088,"Message":"1 validation error detected: Value 'Extreme' at 'quality' failed to satisfy constraint: Member must satisfy enum value set: [High, Normal]","Source":"AudibleApi","TargetSite":"Void ThrowStrongExceptionsIfInvalid(System.String, System.Uri)","RequestUri":"https://api.audible.de/1.0/content/B07SQF3FQ9/licenserequest","JsonMessage":"{\"message\":\"1 validation error detected: Value 'Extreme' at 'quality' failed to satisfy constraint: Member must satisfy enum value set: [High, Normal]\"}","Type":"AudibleApi.ValidationErrorException"},"RequestUri":"/1.0/content/B07SQF3FQ9/licenserequest","JsonMessage":"{\"consumption_type\":\"Download\",\"supported_drm_types\":[\"Adrm\",\"Mpeg\"],\"quality\":\"Extreme\",\"response_groups\":\"last_position_heard,pdf_url,content_reference,chapter_info\"}","Type":"AudibleApi.ApiErrorException"}}2022-07-19 19:40:45.909 +02:00 [ERR] (at LibationWinForms.ProcessQueue.LogMe+<>c.<.ctor>b__9_1(System.Object, System.String)) ERROR. All books have not been processed. Most recent book: processing failed
 {}2022-07-19 19:40:48.988 +02:00 [INF] (at LibationWinForms.ProcessQueue.ProcessQueueControl+<QueueLoop>d__26.MoveNext()) Completed processing queued item: 27/06/2022 [B07SQF3FQ9] Der Dieb und der Söldner: Die Gaukler-Chroniken 1
Result: "FailedRetry"
 {}2022-07-19 19:40:48.994 +02:00 [INF] (at LibationWinForms.ProcessQueue.ProcessQueueControl+<QueueLoop>d__26.MoveNext()) Completed processing queue
```

**Comment by Mbucari** [comment]
@rmcrackan looks like audible just changed their API again.

**Comment by rmcrackan** [comment]
This is weird -- it happened to me too a few hours ago. Then I tried a few minutes later and it worked.

To try again, click the book's right-most button and change it from Error to Not Downloaded. Then try again to download it and let me know if you're still having this error.

**Comment by Mbucari** [comment]
@rmcrackan need to remove Extreme and Low from the enum here:

https://github.com/rmcrackan/AudibleApi/blob/15637064bde881869c0e13f497334cd7ab9f91da/AudibleApi/Api.Download.cs#L12

**Comment by rmcrackan** [comment]
@Mbucari Maybe. Like I said, I had the identical error earlier today. Later I tried to reproduce it so I could find the bug but it worked. I wonder if the books now come with that set of enums and we have to pick the highest one. But that still doesn't explain why one time it failed and one time it didn't.

I'm curious to see what happens when this user tries again.

**Comment by rmcrackan** [comment]
Yeah, now I'm getting your error again with my most recent credit-purchased book. I think @Mbucari is right.

@mkb79 - heads-up: audible might be changing this part of their api. Full error log above (Germany). I'm seeing it too (US)

**Comment by Mbucari** [comment]
https://github.com/mkb79/Audible/discussions/54
I noticed last year that that API endpoint would only ever serve up 2 qualities of audio.

The old Fiona... Sidecar method of downloading (which only worked for aax files, not aaxc or audible plus) would serve 4 qualities of audio. That was the method used by the windows store app. My hunch is now that the windows app is dead, the old Fiona endpoint will be taken away and all audiobook will only be available in 2 qualities: normal and high.

**Comment by rmcrackan** [comment]
> My hunch is now that the windows app is dead...

That makes a lot of sense. I forgot that was at hand.

**Comment by rmcrackan** [comment]
@maaximal -- Much thanks for the details and for your patience. I'll have a fix for you shortly.

**Comment by rmcrackan** [comment]
Fixed in [version 8.2.4](https://github.com/rmcrackan/Libation/releases/latest)
